### PR TITLE
Mon-32851 [gorgone] service auto-discovery only the services associated to one host are created

### PR DESCRIPTION
**Pull request by Evan-Adam**
## Description

since https://github.com/centreon/centreon/pull/1057 some service where not correctly created with the auto discovery.
on average 50% of the service where not created, it's random and can range from 0 to 100% creation in rare case

**Fixes** #MON-32851

## Type of change

- [X] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [X] 23.04.x
- [X] 23.10.x
- [X] 24.04.x (master)

<h2> How this pull request can be tested ? </h2>

Have a central server with the Connector repository and a valid IT license

Install the discovery package and the Linux SNMP connector

`yum install centreon-auto-discovery-server centreon-pack-operatingsystems-linux-snmp centreon-plugin-Operatingsystems-Linux-Snmp`

Install the Linux SNMP connector via the page Configuration  >  Monitoring Connectors Manager

Install the discovery module on the GUI

Create on host Linux1 associated with the OS-Linux-SNMP-custom template. Give it the IP 127.0.0.1

Create on host Linux2 associated with the OS-Linux-SNMP-custom template. Give it the IP address of your central (10.25.XX.XX)

Enable the snmpd service in your central server and make it answer both ip you just configured, for exemple by allowing all ip to connect : 

`agentaddress  0.0.0.0,[::1]`

Enable the rule OS-Linux-SNMP-Disk-Name via the page Configuration  >  Services  >  Rules

push configuration

launch the discovery manually : 

`curl --request POST "http://127.0.0.1:8085/api/centreon/autodiscovery/services" -d '{}'`

check gorgone logs : 

`tail -f /var/log/centreon-gorgone/gorgoned.log`

Go to the page Configuration  >  Services  >  Services by host and check the created services on these hosts

normally each server should have a new service `disk-/` created 
Any **relevant details** of the configuration to perform the test should be added.

#### Community contributors & Centreon team

- [X] I have followed the **coding style guidelines** provided by Centreon
- [X] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [X] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Review by sdepassio — DISMISSED** [review]

**Review by denrou — DISMISSED** [review]

File: centreon-gorgone/docs/modules.md [hunk 1]
```diff
@@ -19,3 +19,57 @@ List of the available modules:
...
   * [Newtest](../docs/modules/plugins/newtest.md)
   * [Scom](../docs/modules/plugins/scom.md)
+
+# Module implementation
+work in progress, may not be complete nor true.
+
+Each module should have a hook.pm and a class.pm file with some mandatory function implemented.
+
+## hook.pm
+mainly used for creating the module process(es)
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules.md:29]:
```suggestion
Mainly used for creating the module process(es)

```

File: centreon-gorgone/docs/modules.md [hunk 2]
```diff
@@ -19,3 +19,57 @@ List of the available modules:
...
+and route event to it each time a new message is received by gorgone.
+
+
+### const EVENTS []
+array defining all events this module can process. Optionally add api endpoint for events.
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules.md:34]:
```suggestion
Array defining all events this module can process. Optionally add API endpoint for events.

```

File: centreon-gorgone/docs/modules.md [hunk 3]
```diff
@@ -19,3 +19,57 @@ List of the available modules:
...
+### const NAME
+### const NAMESPACE
+
+
+### gently()
+Called by gorgone-core when stopping the module.
+### register()
+### init()
+called by library::loadmodule to initialise the module. Should create a child process as it's not made by gorgone-core.
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules.md:43]:
```suggestion
Called by library::loadmodule to initialize the module. Should create a child process as it's not made by gorgone-core.

```

File: centreon-gorgone/docs/modules.md [hunk 4]
```diff
@@ -19,3 +19,57 @@ List of the available modules:
...
+### routing()
+### kill()
+### check()
+### broadcast()
+### create_child()
+Not strictly required, but present every time, used to instantiate a new child process by the init() function.\
+Inside the child process a class.pm object is created and the class->run method is started. 
+
+## class.pm
+this class must inherit module.pm package.\
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules.md:53]:
```suggestion
This class must inherit module.pm package.\

```

File: centreon-gorgone/docs/modules.md [hunk 5]
```diff
@@ -19,3 +19,57 @@ List of the available modules:
...
+this object is most of the time (maybe all ?) a singleton.\
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules.md:54]:
```suggestion
This object is most of the time (maybe all ?) a singleton.\

```

File: centreon-gorgone/docs/modules.md [hunk 6]
```diff
@@ -19,3 +19,57 @@ List of the available modules:
...
+It will be created by hook.pm when starting the module.
+This is the workhorse who will process all event.
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules.md:56]:
```suggestion
This is the workhorse that will process all events.

```

File: centreon-gorgone/docs/modules.md [hunk 7]
```diff
@@ -19,3 +19,57 @@ List of the available modules:
...
+
+It seem like none of theeese method will be called by gorgone-core, so naming are not required to follow this convention
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules.md:58]:
```suggestion
It seems like none of these methods will be called by gorgone-core, so naming is not required to follow this convention.

```

File: centreon-gorgone/docs/modules.md [hunk 8]
```diff
@@ -19,3 +19,57 @@ List of the available modules:
...
+(please keep the code base consistent if you make a new module).
+### new()
+Class constructor
+
+### run()
+will be called by hook.pm. this method should wait for event and dispatch them accordingly.
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules.md:64]:
```suggestion
Will be called by hook.pm. This method should wait for event and dispatch them accordingly.

```

File: centreon-gorgone/docs/modules.md [hunk 9]
```diff
@@ -19,3 +19,57 @@ List of the available modules:
...
+use EV library to wait for new thing to do, either by waiting on the zmq file descriptor (fd)
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules.md:65]:
```suggestion
Uses EV library to wait for new thing to do, either by waiting on the ZMQ file descriptor (fd)

```

File: centreon-gorgone/docs/modules.md [hunk 10]
```diff
@@ -19,3 +19,57 @@ List of the available modules:
...
+It seem like none of theeese method will be called by gorgone-core, so naming are not required to follow this convention
+(please keep the code base consistent if you make a new module).
+### new()
+Class constructor
+
+### run()
+will be called by hook.pm. this method should wait for event and dispatch them accordingly.
+use EV library to wait for new thing to do, either by waiting on the zmq file descriptor (fd) 
+or with a periodic timer.\
+Generally wait for new data on zmq socket with EV::io(), and call event() when there is.
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules.md:67]:
```suggestion
Generally waits for new data on ZMQ socket with EV::io(), and call event() when there is.

```

File: centreon-gorgone/docs/modules.md [hunk 11]
```diff
@@ -19,3 +19,57 @@ List of the available modules:
...
+
+### event()
+read data from zmq socket, and act on it, generally by launching an action_* method to process the event.
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules.md:70]:
```suggestion
Reads data from ZMQ socket, and acts on it, generally by launching an action_* method to process the event.

```

File: centreon-gorgone/docs/modules.md [hunk 12]
```diff
@@ -19,3 +19,57 @@ List of the available modules:
...
+module.pm parent class do have an event() method, so it's not mandatory to implement it.
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules.md:71]:
```suggestion
module.pm parent class has an event() method, so it's not mandatory to implement it. 

```

File: centreon-gorgone/docs/modules.md [hunk 13]
```diff
@@ -19,3 +19,57 @@ List of the available modules:
...
+will be called by hook.pm. this method should wait for event and dispatch them accordingly.
+use EV library to wait for new thing to do, either by waiting on the zmq file descriptor (fd) 
+or with a periodic timer.\
+Generally wait for new data on zmq socket with EV::io(), and call event() when there is.
+
+### event()
+read data from zmq socket, and act on it, generally by launching an action_* method to process the event.
+module.pm parent class do have an event() method, so it's not mandatory to implement it. 
+### action_*()
+method called by event() when a zmq message is found.\
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules.md:73]:
```suggestion
Method called by event() when a ZMQ message is found.\

```

File: centreon-gorgone/docs/modules/centreon/autodiscovery.md [hunk 14]
```diff
@@ -316,3 +316,26 @@ curl --request POST "https://hostname:8443/api/centreon/autodiscovery/services"
     \"dry_run\": 1
 }"
 ```
+
+### Developer manual
+
+This module heavily use the gorgone-action module to work.
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules/centreon/autodiscovery.md:322]:
```suggestion
This module heavily uses the gorgone-action module to work.

```

File: centreon-gorgone/docs/modules/centreon/autodiscovery.md [hunk 15]
```diff
@@ -316,3 +316,26 @@ curl --request POST "https://hostname:8443/api/centreon/autodiscovery/services"
     \"dry_run\": 1
 }"
 ```
+
+### Developer manual
+
+This module heavily use the gorgone-action module to work.
+here is a diagram of how these module interact :
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules/centreon/autodiscovery.md:323]:
```suggestion
Here is a diagram of how these modules interact: 

```

File: centreon-gorgone/docs/modules/centreon/autodiscovery.md [hunk 16]
```diff
@@ -316,3 +316,26 @@ curl --request POST "https://hostname:8443/api/centreon/autodiscovery/services"
...
+
+### Developer manual
+
+This module heavily use the gorgone-action module to work.
+here is a diagram of how these module interact : 
+
+![image](./centreon-gorgone-autodiscovery-archi.png)
+
+
+dotted line mean a ZMQ message is send, direct line mean the function is called normally.
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules/centreon/autodiscovery.md:328]:
```suggestion
Dotted lines mean a ZMQ message is sent. Direct lines mean the function is called normally.

```

File: centreon-gorgone/docs/modules/centreon/autodiscovery.md [hunk 17]
```diff
@@ -316,3 +316,26 @@ curl --request POST "https://hostname:8443/api/centreon/autodiscovery/services"
...
+
+Each column represent a linux thread, as Gorgone is multi-threaded.
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules/centreon/autodiscovery.md:330]:
```suggestion
Each column represents a Linux thread, as Gorgone is multi-threaded.

```

File: centreon-gorgone/docs/modules/centreon/autodiscovery.md [hunk 18]
```diff
@@ -316,3 +316,26 @@ curl --request POST "https://hostname:8443/api/centreon/autodiscovery/services"
...
+
+For each ZMQ message, names are described in the [events section](#events) of each module,
+and for putlog the second part is the 'code' used by gorgone-autodiscovery 
+and defined as constant in the [class.pm](../../../gorgone/modules/centreon/autodiscovery/class.pm) file
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules/centreon/autodiscovery.md:334]:
```suggestion
and defined as constant in the [class.pm](../../../gorgone/modules/centreon/autodiscovery/class.pm) file.

```

File: centreon-gorgone/docs/modules/centreon/autodiscovery.md [hunk 19]
```diff
@@ -316,3 +316,26 @@ curl --request POST "https://hostname:8443/api/centreon/autodiscovery/services"
...
+
+gorgone-action module don't directly send the result to the calling module. Instead it send a putlog message, processed by core.\
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules/centreon/autodiscovery.md:336]:
```suggestion
The gorgone-action module does not send the result directly to the calling module. It sends a putlog message instead, processed by core.\
```

File: centreon-gorgone/docs/modules/centreon/autodiscovery.md [hunk 20]
```diff
@@ -316,3 +316,26 @@ curl --request POST "https://hostname:8443/api/centreon/autodiscovery/services"
...
+core keep track  of every module waiting for a particular event (use library.pm::addlistener to show interest in an event)\
```
Comment (smau-centreon) [inline centreon-gorgone/docs/modules/centreon/autodiscovery.md:337]:
```suggestion
Core keeps track of every module waiting for a particular event (using library.pm::addlistener to show interest in an event)\

```

**Review by sdepassio — APPROVED** [review]

**Review by denrou — APPROVED** [review]

**Comment by cgagnaire** [comment]
Hi,
I witness the same behaviour when more than 1 discovery is done at the same time.
It works if you put service_parrallel_commands_poller to 1 and launch the discovery on only 1 Poller.
So, yes some listeners get lost, but this PR looks like a wild guess and not a strong fix.
Also the lines touched with no valid reason makes it not very readable.
I got in touch with @garnier-quentin and we will look at it next week, so if you're willing to wait a bit before merging I think he could provide a better fix.

**Pull request merged**